torch/csrc/jit/passes/peephole.cpp

```diff
@@ -303,6 +303,15 @@ struct PeepholeOptimizeImpl {
               output->debugName());
           node->output()->replaceAllUsesWith(output);
         }
+      } else if (node->kind() == prim::profile) {
+        auto inp = node->input(0)->node();
+        // we remove back-to-back prim::profiles if a consumer
+        // immediately follows a producer and their profiled types are equal
+        if (inp->kind() == prim::profile && node->prev() == inp &&
+            inp->ty(attr::profiled_type) == node->ty(attr::profiled_type)) {
```

Comment:
How could we end-up with two consecutive prim::profile nodes with different types? Could we assert instead?

Reply:
If there's any optimization I think it's possible, it's hard to come up with a case but i dont think we should assert either.

Reply:
An ultimate solution would be, imho, to 1) assert that adjacent profile nodes don't contradict each other, 2) find an intersection of the profiled info for the profiled nodes and replace them with a single node.
E.g.
```
%y = prim::profile[type=Float(2,3)](%x)
%z = prim::profile[type=Float(10,3)](%y)
```
should trigger an error, but
```
%y = prim::profile[type=Float(*,*, device=cpu)](%x)
%z = prim::profile[type=Float(10,3)](%y)
```
should be optimized to
```
%z = prim::profile[type=Float(10,3,device=cpu)](%x)
```

Reply:
But I think having a partial solution would suffice for now, so the current approach looks good to me.